### PR TITLE
Feature: cashflow report by service

### LINKED ISSUE #165: (feature request) Report of Cashflow by Service
The Vanga Hospital tracks much of its income by service.  At the end of the day, they would like to know the amount of money in the cashbox realized by service at the end of the day.  This involves a LEFT JOIN between the invoices and cash payments on all cash payments realized that day.  The breakdown should look something like this:

| Invoice | Debtor |  Payment | Service A | Service B | ... | Service Z |
| --- | --- | --- | --- | --- | --- | --- |
| IV.HEV.12 | PA.HEV.34 | $45.23 |  -  |  $45.23 | ... | - |
| IV.HEV.45 | PA.HEV.23 | $3.01 | $3.01  | - | ... | - |
| IV.HEV.20 | PA.HEV.19 | $7.93 | $7.93  | - | ... | - |
| IV.HEV.14 | PA.HEV.36 | $0.54 | -  | - | ... | $0.54 |
|  **GRAND TOTAL** |  | **$56.71** | **$10.94** | **$45.23** | ... | **$0.54** |

### PR DESCRIPTION
**Pull request by jniles**
This pull request implements the server-side of the cashflow-by-service report.  The client-side is not yet developed, but will use the same report creation/saving infrastructure as all other reports.  This PR is submitted mainly for review before dashing forward with the rest of the app.

For reference, here is the output:
cashflowbyservice
_Fig 1: Cashflow by Service Report_

If this report's logic looks fine, the next steps are:
 1. ~~Translate all keys~~
 2. Write more complex filters - I believe a tool to help this was started a little over a week ago.
 3. ~~Create the client side.~~
 4. ~~Write integration tests for the PDF route.~~

Closes #165.

---------
Thank you for contributing!

Before submitting this pull request, please verify that you have:
 - [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
 - [x] Run integration tests.
 - [x] Run end-to-end tests.
 - [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

### CODE REVIEW

**Comment by jniles** [comment]
@sfount, all portions of this PR are implemented.  You can do a full review now!

**Comment by sfount** [comment]
I'll look through this now.

File: server/controllers/finance/reports/cashflow/index.js [hunk 1]
```diff
@@ -564,3 +570,133 @@ function document(req, res, next) {
...
+
+      // map services to their service names
+      data.services = services.map(service => service.name);
+
+      let xAxis = data.services.length;
+
+      // file the matrix with nulls except the correct columns
+      let matrix = rows.map(row => {
+
+        // fill with each service + two lines for cash payment identifier and patient name
```
Comment (sfount) [inline server/controllers/finance/reports/cashflow/index.js]: Very useful comments.

File: client/src/partials/reports/modals/cashflowByService.config.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+CashflowByServiceConfigController.$inject = [
+  '$state', '$http', '$uibModalInstance', 'NotifyService', 'LanguageService', 'reportDetails'
+];
+
+/**
+ * CashflowByService
+ *
+ * @description
+ * This controller is responsible of Aged Debtors report, that report include
```
Comment (sfount) [inline client/src/partials/reports/modals/cashflowByService.config.js]: As it has been pointed out by others *who are definitely not you*, this description should reflect the cash flow report that the method represents.

**Comment by sfount** [comment]
Issue on testing system reports https://github.com/Vanga-Hospital/bhima-2.X/issues/228

**Comment by jniles** [comment]
@sfount, changes made!

**Pull request merged**